docker-qgis/process_projectfile.py

```diff
@@ -1,40 +1,68 @@
 from pathlib import Path
-from typing import Dict
+from typing import NamedTuple, Optional
 from xml.etree import ElementTree
 
-from qfieldcloud.qgis.utils import BaseException, get_layers_data, layers_data_to_string
+from qfieldcloud.qgis.utils import (
+    FailedThumbnailGenerationException,
+    InvalidFileExtensionException,
+    InvalidXmlFileException,
+    ProjectFileNotFoundException,
+    get_layers_data,
+    layers_data_to_string,
+)
 from qgis.core import QgsMapRendererParallelJob, QgsMapSettings, QgsProject
 from qgis.PyQt.QtCore import QEventLoop, QSize
 from qgis.PyQt.QtGui import QColor
 
 logger = logging.getLogger("PROCPRJ")
 
 
-class ProjectFileNotFoundException(BaseException):
-    message = 'Project file "%(project_filename)s" does not exist'
+class XmlErrorLocation(NamedTuple):
+    line: int
+    column: int
 
 
-class InvalidFileExtensionException(BaseException):
-    message = (
-        'Project file "%(project_filename)s" has unknown file extension "%(extension)s"'
-    )
-
-
-class InvalidXmlFileException(BaseException):
-    message = "Project file is an invalid XML document:\n%(xml_error)s"
+def get_location(invalid_token_error_msg: str) -> Optional[XmlErrorLocation]:
+    """Get column and line numbers from the provided error message."""
+    if "invalid token" not in invalid_token_error_msg.casefold():
+        logger.error("Unable to find 'invalid token' details in the given message")
+        return None
 
+    _, details = invalid_token_error_msg.split(":")
+    line, column = details.split(",")
+    _, line_number = line.strip().split(" ")
+    _, column_number = column.strip().split(" ")
 
-class InvalidQgisFileException(BaseException):
-    message = 'Project file "%(project_filename)s" is invalid QGIS file:\n%(error)s'
+    return XmlErrorLocation(int(line_number), int(column_number))
 
 
-class InvalidLayersException(BaseException):
-    message = 'Project file "%(project_filename)s" contains invalid layers'
+def contextualize(
+    invalid_token_error_msg: str, fh: io.BufferedReader
+) -> Optional[tuple[str, str, str]]:
+    """
+    Get an html-safe slice of the line where the exception occurred, with all faulty occurrences sanitized.
+    Makes no use of '.decode(..., errors="replace")' because it still throws on some entities.
+    """
+    location = get_location(invalid_token_error_msg)
+    if location:
+        substitute = "?"
+        fh.seek(0)
+        for cursor_pos, line in enumerate(fh, start=1):
+            if location.line == cursor_pos:
+                faulty_char = line[location.column]
+                suffix_slice = line[: location.column - 1]
+                clean_safe_slice = suffix_slice.decode("utf-8").strip() + substitute
 
+                return (
+                    f"Unable to parse this character: {repr(faulty_char)}",
+                    f"It was replaced by '{substitute}' on line {location.line} that starts with:",
+                    html.escape(clean_safe_slice),
```

Comment:
The escape should not be here. The HTML injection is a visualization problem, not a storage problem.

Reply:
I do agree it is a quick solution, but I think it is quite dirty. Since we do direct queries in jobs to search for errors in the database, escaping the output in such a early stage will make our task much harder, as we have to remember the specific HTML escaping.

What is more, what if we want to put the outputs in CSV one day, should we already escape? Or JSON, yaml etc etc.